[Task] Add early linting to the Swift test

JJ opened the pull request:
Linting using `swiftlint` as suggested by @GhenadiePusca, and using a Docker container, which seems the fastest and does not need additional local configuration.
Still, initial running returned a lot of warnings/errors, so these might have to be fixed before merging (if approved)

JJ commented:
As expected there's a ton of linting errors. Won't work on them (either disabling or editing source) until we're really sure we want this.

GhenadiePusca commented:
> As expected there's a ton of linting errors. Won't work on them (either disabling or editing source) until we're really sure we want this.

Took a quick look on the lint errors, most of them make sense. I can take care of them.

JJ commented:
> > As expected there's a ton of linting errors. Won't work on them (either disabling or editing source) until we're really sure we want this.
> 
> Took a quick look on the lint errors, most of them make sense. I can take care of them.

From my POV, this just adds a few seconds to the overall workflow, and only when it's needed. So I think we should have it. I can help with the linting errors, also, if you want, but please wait a bit until we're sure @fhd is on board with this.

fhd commented:
I am!

**But:**

A) I think our strategy should be to enforce pretty much only our _existing_ rules right now, not change everything, basically change as little as possible in this step. Examples:

1. Disable the cyclomatic complexity stuff
2. Disable the trailing whitespace rule (that's how Xcode auto indents - I'm only on board with changing that in the linter if we manage to find a way to change it in Xcode, but I think this should be a **separate** PR)

The other warnings look pretty legit from what I've seen, but overall I would still lean towards _disabling_ checks now, finishing this PR, and then enabling rules we want to have again PR-by-PR.

B) _If_ any code needs to be changed, I think we should land this one only after #748 has landed - that poor PR has been fighting an uphill battle against all sorts of refactorings.

JJ commented:
After disabling the two indicated, there's still a sh*tload of them. I'll disable most of them, except maybe some that are simple enough that can be fixed on the fly.

JJ commented:
I've seen there's a rule that flags "TODOs". I should *leave* that one.

GhenadiePusca commented:
I like @fhd suggestion! Maybe, @JJ , with this PR you can choose to fix one kind of SwiftLint error, like trailing space and disable the rest. And then with consequent PRs  we will enable rules one by one, until we have the full set.

JJ commented:
I've fixed the ones that were less frequent, or touching the least files, like trailing semicolon. While I was doing those, I've also fixed trailing whitespace, but it's still disabled for most.

GhenadiePusca commented on platform/ios/.swiftlint.yml:
```diff
@@ -0,0 +1,24 @@
+disabled_rules:
```
I think it is possible to use `only_rules` to specify which rules should be enabled, instead of saying which rules should be disabled.

JJ replied:
I preferred to take this approach, so that we can later eliminate them one by one (or two by two) in separate PRs... TBH. But if you think that's a better approach, I pretty much know which rules I fixed in this PR, I can put them there if needed

GhenadiePusca replied:
Well I don't know, yet, if we should use all of the SwiftLint rules. And `only_rules` is explicit as to what rules we are using, instead of what we are not using.

How is it for the other languages, like JS, Kotlin, TS, are we using all of the default lint rules so we follow community standards? Or we imply some of our own reasoning/liking as to what rules should be added? It will be nice to be consistent with other languages.

JJ replied:
Generally, yes. We pretty much worked with Kotlin in the same way. ES was already here when I arrived, though, but looks like it was done in the same way (at least looking at how we've disabled them recently.

GhenadiePusca replied:
Ok then, we are good to go with all the rules. And disable nasty ones when needed.

GhenadiePusca commented on .github/workflows/js-and-android-test.yml:
```diff
@@ -40,9 +40,15 @@ jobs:
             - name: Lint core and features
               if: steps.changes.outputs.js == 'true'
               run: ./build.js lint
+            - name: Lint Swift
```
Not sure if we want to do it in this PR, but also the name of the workflow file needs to be updated to include also iOS.

JJ replied:
Not really doing any iOS stuff, other than linting Swift... But we might want to add "linting" to the name... somewhere (kotlint is done elsewhere, for no reason)

GhenadiePusca replied:
No strong preference. I was mostly thinking of bundling related jobs in the same workflow. But you know more about the whole CI setup, so if it is better to have it running on Ubuntu let's keep it as is.

GhenadiePusca reviewed (approved):
Awesome start 💯!

GhenadiePusca commented:
> It's still removing trailing whitespace, my point is that we can **not** do that while Xcode basically _automatically indents_ code that way. It's way too frustrating to have Xcode follow one style, and swiftlint follow a _competing_ style. They need to be the same, I don't care so much about which one it is, but they cannot be different.

It is mostly due to Xcode config -> https://stackoverflow.com/questions/1390329/trim-trailing-spaces-in-xcode

fhd commented:
I know, but it's pretty impractical to ask everyone working on this code to manually fiddle with their Xcode settings. I'm sure there's a way to set it via the repository somehow.

But I also think that should not block progress on the otherwise fantastic endeavour of adding swiftlint! I'm merely arguing that we shouldn't introduce any rules that are the **opposite** of what everyone's IDE currently does.

GhenadiePusca commented:
> I know, but it's pretty impractical to ask everyone working on this code to manually fiddle with their Xcode settings. I'm sure there's a way to set it via the repository somehow.
> 
> But I also think that should not block progress on the otherwise fantastic endeavour of adding swiftlint! I'm merely arguing that we shouldn't introduce any rules that are the **opposite** of what everyone's IDE currently does.
> 
> To be clear, my point is that **Xcode actively inserts trailing whitespace when you reindent a file, or generate a new one** - it's not just that it doesn't remove trailing whitespace automatically, that'd be fine. My issue is really that it actively _inserts_ it.

The whitespace on a new line(even if it is empty) is added to preserve indentation. iOS developers usually enable the functionality of removing that whitespace. I have worked on big teams, where initial config was to have that checkmark enabled. The main concern are merge conflicts, when an empty line was changed to contain code. Of course, one can use `--ignore-space-change` to maybe mitigate this.

fhd commented:
Well, if swiftlint won't complain about the trailing whitespace, that's better.

But still, in that sense, this PR contains _unrelated_ changes of removing trailing whitespace on empty lines - which is not in line with how we currently do things, people actively run `^i` on all Swift files they touch, which adds this trailing whitespace back.

We _can_ block swiftlint until we've figured out how to change everybody's setting etc, I'd just rather tackle that issue separately.

JJ commented:
> Well, if swiftlint won't complain about the trailing whitespace, that's better.
> 
> But still, in that sense, this PR contains _unrelated_ changes of removing trailing whitespace on empty lines - which is not in line with how we currently do things, people actively run `^i` on all Swift files they touch, which adds this trailing whitespace back.
> 

The problem is that in many cases there are several blocks of whitespace. It's not a big deal if people add them back. It *is* a big deal if I have to put those blocks of whitespace back, since I would have to restore the original version, re-apply non-whitespace-related changes, and submit back. To be clear, I've removed only whitespace in files where there were other changes to do related to the rules that have been left in place.

> We _can_ block swiftlint until we've figured out how to change everybody's setting etc, I'd just rather tackle that issue separately.

No need to do that. As I say, if people edit those files for any reason and whitespace is added, no one is going to complain, neither swiftlint nor me nor anyone.

GhenadiePusca commented:
About the whole topic of whitespaces on an empty line, it is not just about SwiftLint. I mean the community standard is to not have whitespaces on the empty lines, even Apple own SDK's comply to it. So how Xcode behaves is more like a "bug".

We can decide in a separate PR if we want to comply to this.

JJ commented:
> About the whole topic of whitespaces on an empty line, it is not just about SwiftLint. I mean the community standard is to not have whitespaces on the empty lines, even Apple own SDK's comply to it. So how Xcode behaves is more like a "bug".
> 
> We can decide in a separate PR if we want to comply to this.

Be that as it may, my point here is that it's going to be more difficult for me to add back those (ugly) whitespaces in current, modified files (which were *only* modified because they had other issues) than for whoever wants them back to add them if they edit these specific files.

Even more so, when we decide to eliminate whitespace in Swift, which is the standard, these files will be already OK.

GhenadiePusca commented:
Recovered most of the removed whitespaces on the newlines. Some where reduced to normal spacing. I see that the codebase is inconsistent, in the same file there are newlines with and without whitespaces.

GhenadiePusca commented on platform/ios/PolyPodApp/FeatureContainer/FeatureContainerView.swift:
```diff
@@ -83,7 +83,7 @@ class FeatureFileHandler: UIViewController, WKURLSchemeHandler {
         let index = urlString.index(urlString.startIndex, offsetBy: PolyOut.fsPrefix.count)
         let file = String(urlString[index..<urlString.endIndex]).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
         let ext = (file as NSString).pathExtension
-                        
```
This is like 16 spaces... removed.

fhd reviewed (approved):
LGTM!

The whitespace that's being removed now is actually nothing that Xcode would produce, it just adds trailing whitespace to empty lines up to the current indentation level.

fhd commented:
@GhenadiePusca Yes, it is a bit inconsistent, but at the moment we (A) leave all Xcode settings at the default and (B) run _CTRL_ + _i_ on every Swift file we touch. I think we have two ways towards getting this consistent now:

1. Changing the Xcode settings to not produce trailing whitespace (or at _least_ documenting that everybody needs to change it), then re-adding the trailing whitespace rule and fixing all files
2. Manually opening each Swift file and hitting _CMD_ + _a_ followed by _CTRL_ + _i+.

I'm on board with (1) - just in a separate PR is all I ask :D
